core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -598,6 +603,86 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
     }, "Broker fail to initialize after restart")
   }
 
+  @Test
+  def testPreemptionOnControllerShutdown(): Unit = {
+    servers = makeServers(1, enableControlledShutdown = false)
+    val controller = getController().kafkaController
+    val count = new AtomicInteger(2)
+    val latch = new CountDownLatch(1)
+    val spyThread = spy(controller.eventManager.thread)
+    controller.eventManager.setControllerEventThread(spyThread)
+    val processedEvent = new MockEvent(ControllerState.TopicChange) {
+      override def process(): Unit = latch.await()
+      override def preempt(): Unit = {}
+    }
+    val preemptedEvent = new MockEvent(ControllerState.TopicChange) {
+      override def process(): Unit = {}
+      override def preempt(): Unit = count.decrementAndGet()
+    }
+
+    controller.eventManager.put(processedEvent)
+    controller.eventManager.put(preemptedEvent)
+    controller.eventManager.put(preemptedEvent)
+
+    doAnswer((_: InvocationOnMock) => {
+      latch.countDown()
+    }).doCallRealMethod().when(spyThread).awaitShutdown()
```

Comment:
@jsancio In the suggested code, `latch.await()` blocks the event thread from shutting down while `thread.awaitShutdown()` (called in `controller.shutdown()`) blocks the main thread from reaching `latch.countDown()` causing a deadlock. The fix is to call `latch.countDown()` before `thread.awaitShutdown()` which has been implemented using `spy`.

Please let me know if there's another way to keep the `thread` field as a `val`.

Reply:
Could we have the test call `latch.countDown()` in a background thread with a delay, right before we call controller.shutdown? I guess it's possible to have race conditions with that solution

Reply:
@stanislavkozlovski thanks for the comment. i've tried this approach before and the test passes but sometimes output:
```
[2020-07-23 11:12:43,316] ERROR [RequestSendThread controllerId=0] Controller 0 fails to send a request to broker localhost:51542 (id: 0 rack: null) (kafka.controller.RequestSendThread:76)
java.lang.InterruptedException
```
i also see a bit of flakiness with this approach as we cannot exactly time when `latch.countDown()` is called.

Reply:
Got it. Sounds good.

Reply:
The important part is that we want to do `latch.countDown` after `initiateShutdown` has been called so that the controller thread doesn't pick up a new event because `isRunning` is `false`.